fitz/utils.py

```diff
@@ -1834,6 +1834,141 @@ def insert_text(
     return rc
 
 
+def insert_htmlbox(
+    page,
+    rect,
+    text,
+    *,
+    css=None,
+    adjust=True,
+    archive=None,
+    rotate=0,
+    oc=0,
+    overlay=True,
+    morph=None,
+) -> float:
+    """Insert text with optional HTML tags and stylings into a rectangle.
+
+    Args:
+        rect: (rect-like) rectangle into which the text should be placed.
+        text: (str) text with optional HTML tags and stylings.
+        css: (str) CSS styling commands.
+        adjust: (bool) force-fit the text by scaling down (default True).
+        archive: Archive object pointing to locations of used fonts or images
+        rotate: (int) rotate the text with the box by a multiple of 90 degrees.
+        oc: (int) the xref of an OCG / OCMD (Optional Content).
+        overlay: (bool) put text on top of page content.
+        morph: (sequence) (point, matrix) transform result with a matrix
+               using a fixed point.
+    Returns:
+        A float indicating the unused part of the rectangle. If negative, the
+        rectangle was too small for the text.
+    """
+
+    def story_maker(clip, text, adjust=True, rotate=0, css=css, archive=archive):
+        links = []
+
+        def collect(elpos):
+            if not elpos.open_close & 2 or elpos.href is None:
+                return
+            links.append((elpos.href, elpos.rect))
+            return
+
+        rect = Rect(clip)  # copy to rect
+        # shift to top-left = (0, 0)
+        rect += (-clip.x0, -clip.y0, -clip.x0, -clip.y0)
+        if rotate in (90, 270):  # swap width height
+            rect.x1, rect.y1 = rect.y1, rect.x1
+        orig_height = rect.y1  # original relevant height
+        mycss = "body {margin:1px;}" + css  # append user CSS
+        if isinstance(text, str):  # if a string, convert to a Story
+            story = Story(html=text, user_css=mycss, archive=archive)
+        else:
+            story = text  # assume we have a story
+
+        # ---------------------------------------------------------------------
+        # Loop for fitting text into rectangle.
+        # We will expand the rectangle until we have a fit in any case. Only
+        # finally we will check "adjust" and report the overflow amount if
+        # no adjustment was intended.
+        # ---------------------------------------------------------------------
+        more = True
+        while more:
+            story.reset()  # rewind story to the beginning
+            fp = io.BytesIO()  # use for file output
+            writer = DocumentWriter(fp)
+            dev = writer.begin_page(rect)  # make a temporary page
+            more, filled = story.place(rect)  # position text in the rect
+            if more:  # text did not fit in this clip, so enlarge and try again
+                rect *= 1.01  # enlarge by 1%
+                links = []
+                continue
```

Comment:
If `adjust` is false, should we still be enlarging the rect here?

Reply:
Yes - to enable us computing the (approximate) overflow value.